src/afl-fuzz-queue.c

```diff
@@ -860,3 +861,130 @@ u32 calculate_score(afl_state_t *afl, struct queue_entry *q) {
 
 }
 
+void queue_testcase_retake(afl_state_t *afl, struct queue_entry *q,
+                           u32 old_len) {
+
+  if (q->testcase_buf) {
+
+    munmap(q->testcase_buf, old_len);
+    int fd = open(q->fname, O_RDONLY);
+
+    if (unlikely(fd < 0)) { PFATAL("Unable to open '%s'", q->fname); }
+
+    u32 len = q->len;
+    q->testcase_buf = mmap(0, len, PROT_READ | PROT_WRITE, MAP_PRIVATE, fd, 0);
+
+    if (unlikely(q->testcase_buf == MAP_FAILED)) {
+
+      PFATAL("Unable to mmap '%s' with len %d", q->fname, len);
+
+    }
+
+    close(fd);
+    afl->q_testcase_cache_size = afl->q_testcase_cache_size + q->len - old_len;
+
+  }
+
+}
+
+/* Returns the testcase buf from the file behind this queue entry.
+  Increases the refcount. */
+inline u8 *queue_testcase_get(afl_state_t *afl, struct queue_entry *q) {
+
+  u32 len = q->len;
+
+  /* first handle if no testcase cache is configured */
+
+  if (likely(!afl->q_testcase_cache_size)) {
```

Comment:
see my other comment. a cache that only fits < 50% of the testcases hurts performance because of the evictions. so a default cache is not a good idea IMHO. and we need code for the option that someone wants that disabled (e.g. fuzzing on embedded that does not have a lot of ram). so thats why I propose to leave the default to no cache. however if none is set a comment about this is printed on startup.

Reply:
I disagree, but we should maybe test it: Evictions are quasi free, a few extra isns, no runs to the kernel.

Reply:
yeah lets test this on fuzzbench after this PR has landed. one variant with 0MB, one with 1MB and another with 1024MB.